File: algorithms/math/lucas-theorem/python/lucas_theorem.py
def lucas_theorem(n, k, p):
    if k > n:
        return 0

    # Precompute factorials mod p
    fact = [1] * p
    for i in range(1, p):
        fact[i] = fact[i - 1] * i % p

    def mod_inv(a, m):
        return pow(a, m - 2, m)

    def comb_small(a, b):
        if b > a:
            return 0
        if b == 0 or a == b:
            return 1
        return fact[a] * mod_inv(fact[b], p) % p * mod_inv(fact[a - b], p) % p

    result = 1
    while n > 0 or k > 0:
        ni = n % p
        ki = k % p
        if ki > ni:
            return 0
        result = result * comb_small(ni, ki) % p
        n //= p
        k //= p

    return result


if __name__ == "__main__":
    print(lucas_theorem(10, 3, 7))
    print(lucas_theorem(5, 2, 3))
    print(lucas_theorem(100, 50, 13))
    print(lucas_theorem(3, 5, 7))
    print(lucas_theorem(0, 0, 5))

File: algorithms/math/lucas-theorem/rust/lucas_theorem.rs
fn mod_pow(mut base: i64, mut exp: i64, m: i64) -> i64 {
    let mut result = 1i64; base %= m;
    while exp > 0 {
        if exp & 1 == 1 { result = result * base % m; }
        exp >>= 1; base = base * base % m;
    }
    result
}

fn lucas_theorem(mut n: i64, mut k: i64, p: i64) -> i64 {
    if k > n { return 0; }
    let mut fact = vec![1i64; p as usize];
    for i in 1..p as usize { fact[i] = fact[i - 1] * i as i64 % p; }

    let mut result = 1i64;
    while n > 0 || k > 0 {
        let ni = (n % p) as usize;
        let ki = (k % p) as usize;
        if ki > ni { return 0; }
        let c = fact[ni] * mod_pow(fact[ki], p - 2, p) % p * mod_pow(fact[ni - ki], p - 2, p) % p;
        result = result * c % p;
        n /= p; k /= p;
    }
    result
}

fn main() {
    println!("{}", lucas_theorem(10, 3, 7));
    println!("{}", lucas_theorem(5, 2, 3));
    println!("{}", lucas_theorem(100, 50, 13));
}
